### ISSUE TITLE
Missing validation for ambiguous constructor call

### ISSUE THREAD

**Issue by szarnekow** [eclipse/xtext-extras#389]
`new FileInputStream(null)` should yield an error marker but does not.

**Comment by szarnekow** [comment]
`System.out.println(null)` is flagged with an error though.

**Comment by szarnekow** [comment]
`new TreeSet(null)` is not ... probably only buggy for constructors?

**Comment by cdietrich** [comment]
see 
https://bugs.eclipse.org/bugs/show_bug.cgi?id=474965 
https://bugs.eclipse.org/bugs/show_bug.cgi?id=461325
too

**Comment by cdietrich** [comment]
the comparison of the candidates returns `EQUALLY_INVALID` and not `AMBIGUOUS`
(coming from EQUALLY_INVALID in compareByArgumentTypes
so `org.eclipse.xtext.xbase.typesystem.internal.AbstractPendingLinkingCandidate.getPreferredCandidate(ILinkingCandidate)` returns the first one and not a `AmbiguousLinkingCandidate`

i have no idea why it is done that way.

**Comment by cdietrich** [comment]
difference seems to be in `org.eclipse.xtext.xbase.typesystem.internal.AbstractPendingLinkingCandidate.compareByArgumentTypesFlags(AbstractPendingLinkingCandidate<?>, int, int, int, int)`
but that foes too deep into xbase knowledge for me
cc @szarnekow

**Comment by szarnekow** [comment]
Looks like we are missing an override in the ConstructorLinkingCandidate similar to the TypeLiteralLinkingCandidate.
```
	protected int getConformanceFlags(int idx, boolean recompute) {
		if (idx == -1) {
			return ConformanceFlags.CHECKED_SUCCESS;
		}
		return super.getConformanceFlags(idx, recompute);
	}
```

If constructors are queried for the conformance of their receiver, they should return a success value instead of an invalid value. Two equally invalid constructor candidates will be chosen arbitrarily by the type computer. If they are both valid, they will be compared more thoroughly.

**Comment by szarnekow** [comment]
```
	def void m() {
		new C(null) 
	}
	
	static class C {
		new(String s) {}
		new(Void v) {}
	}
```
will correctly yield 
```
Ambiguous constructor call.
The constructors
	C(String) and
	C(Void)
both match.
```

**Comment by cdietrich** [comment]
Fixed in 2.19. Thx @szarnekow
